README.md
# PD_fingertip_detection
Codes for finding coordinates of thumb & fingertip of hand across images 

### Train-test-validation split
script_for_train_test_valid_pics_split.ipynb : Script splits the videos into frames and shuffles them across train-test-validation folders

### Fingertip detection

#### Pre-process folder
augmentation.py : Contains all the data augmentation steps
generator.py : Script for generation of training and validation images & ground truth labels for training the models. 

#### Utils folder
history.py: Code for generating the loss-epoch curve 

#### Net folder
Contains codes for calling different model architectures along with the configuration of top FC layers. 

#### Weights
Folder for saving the trained model weights.

#### Training the model
train.py: Code for initiating the model training. 
hyperparam_train.py : Code for hyperparameter tuning with early stopping.
performance.py: Code for evaluating the performance o the trained model.
Script for plotting & checking predicted labels.ipynb : Code for using the trained model to predict and plot the fingertip coordinates on any image, alongwith the ground truth labels. 



history.txt
{'val_loss': [0.002232034360834708, 0.0018123018235201016, 0.0017681423632893711, 0.0012967610964551568, 0.0014710763692467783, 0.0012001804522393893, 0.0012483710476468939, 0.0012097106518922374, 0.0013731676469130132, 0.0010747755271343824, 0.0014123045742356528, 0.0009321796484679604, 0.001331727309055471, 0.0008953867121211564, 0.0010799869778566062, 0.0009473518657614477, 0.00110948137929275, 0.0006977645073978541, 0.0011549893921862047, 0.0008196827475330792, 0.0009682956151664257, 0.000812846667637738, 0.0010128430391584213, 0.0007231765169611511, 0.0009742741700999128, 0.0007635878913182145, 0.0009424639792996459, 0.0006843124744288313, 0.0009389455929825393, 0.0007438966276822612, 0.0008926884547690861, 0.0007861018748371862, 0.0007919733034214005, 0.0007617542141815647, 0.0008213343996127757, 0.0006763684468751308, 0.0007594403517335498, 0.0007849034312433408, 0.0005852129021756506, 0.0007799957093084231, 0.0006762978227925487, 0.0008034667504640917, 0.0006031733440371075, 0.0007426634074363392], 'loss': [0.018362477462002893, 0.006034332106930468, 0.004076904091225179, 0.003206053285221397, 0.002736098287170063, 0.0023600124275415307, 0.002182578579608903, 0.0019473812882759734, 0.0018377183263914452, 0.0016884089312458664, 0.0016492567658223166, 0.001514465530851373, 0.0014479364560238649, 0.0013638217617286208, 0.0013186069876228654, 0.0012407852318799008, 0.001222569781267889, 0.0011236869631605338, 0.0011247235217726687, 0.0010258560121004802, 0.0010387146010790826, 0.0009556534029270327, 0.0009762657286001392, 0.0009015650507479469, 0.0009420234082988176, 0.0008612941290513486, 0.0008897559839821087, 0.0008308883836524515, 0.0008790421514860986, 0.0008213648509326172, 0.0008690008503042828, 0.0007948572919876487, 0.0008487287451754566, 0.0007666820476268545, 0.0008112053143616512, 0.0007431585107064794, 0.000783490745550493, 0.0007200730985381215, 0.0007828904697053526, 0.0007194794718403208, 0.000755302501299108, 0.0006881883791159195, 0.0007555592211572753, 0.0006795479367375029]}


test_colab.py
print("this file could run")

test_predict.py
import cv2
from yolo.detector import YOLO
from fingertip import Fingertips

hand = YOLO(weights='weights/yolo.h5', threshold=0.5)
fingertip = Fingertips(model='vgg', weights='weights/vgg16.h5')

image_file = 'img037.png'
image = cv2.imread(image_file)
tl, br = hand.detect(image=image)

if tl and br is not None:
    alpha_x = 0
    alpha_y = 0
    tl = (tl[0] - alpha_x, tl[1] - alpha_y)
    br = (br[0] + alpha_x, br[1] + alpha_y)

    xmin = int(tl[0])
    ymin = int(tl[1])
    xmax = int(br[0])
    ymax = int(br[1])

    ymin = ymin if ymin > 0 else 0
    xmin = xmin if xmin > 0 else 0

    cropped_image = image[ymin:ymax, xmin:xmax]
    cols, rows, _ = cropped_image.shape
    cropped_image = cv2.resize(cropped_image, (128, 128))
    # cv2.imwrite('cropped_image.jpg', cropped_image)
    position = fingertip.classify(image=cropped_image)

    position[0] = (position[0]) * rows + tl[0]
    position[1] = (position[1]) * cols + tl[1]
    position[2] = (position[2]) * rows + tl[0]
    position[3] = (position[3]) * cols + tl[1]

    """ Drawing bounding box and fingertip """
    image = cv2.rectangle(image, tl, br, (235, 26, 158), 4, 1)
    image = cv2.circle(image, (int(position[0]), int(position[1])), 14, (0, 0, 255), -1)
    image = cv2.circle(image, (int(position[2]), int(position[3])), 14, (0, 255, 0), -1)
    cv2.imshow('predicted_image', image)
    cv2.imwrite('data/' + image_file[:-4] + '_out.jpg', image)
    cv2.waitKey(0)


train.py
import tensorflow as tf
from keras.optimizers import Adam
from keras.callbacks import ModelCheckpoint
from net.vgg16 import model
from preprocess.generator import train_generator, valid_generator
from fingertip import Fingertips ## Kanav adding this so that we can use transfer learning

#import os
#os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'

def loss_function(y_true, y_pred):
    square_diff = tf.squared_difference(y_true, y_pred)
    square_diff = tf.reduce_mean(square_diff, 1)
    loss = tf.reduce_mean(square_diff)
    return loss


# creating the model
model = model()
#model.load_weights('weights/vgg16.h5') # Kanav added this so we can use transfer learning.
model.summary()

# compile
adam = Adam(lr=1e-2, beta_1=0.9, beta_2=0.999, epsilon=1e-10, decay=0.0) #changing lr to 1e-2 from 1e-5
model.compile(optimizer=adam, loss=loss_function)

# train
epochs = 2 # kanav changed from 10 to 2
train_gen = train_generator(sample_per_batch=1, batch_number=8232) # kanav changed from 1690 to 5, sample_per_batch to 1
val_gen = valid_generator(sample_per_batch=3, batch_number=2) # kanav changed sample_per_batch from 50 to 3 & batch number from 10 to 2 

checkpoints = ModelCheckpoint('weights/weights{epoch:03d}.h5', save_weights_only=True, period=1)
history = model.fit_generator(train_gen, steps_per_epoch=8232, epochs=epochs, verbose=1, shuffle=True,  # kanav changed from 1690 to 3
                              validation_data=val_gen, validation_steps=2, #changed from 10 to 2 - nto sure of original though
                              callbacks=[checkpoints], max_queue_size=100)

with open('history.txt', 'a+') as f:
    print(history.history, file=f)

print('All Done!')


train-orig.py
import tensorflow as tf
from keras.optimizers import Adam
from keras.callbacks import ModelCheckpoint
from net.vgg16 import model
from preprocess.generator import train_generator, valid_generator


def loss_function(y_true, y_pred):
    square_diff = tf.squared_difference(y_true, y_pred)
    square_diff = tf.reduce_mean(square_diff, 1)
    loss = tf.reduce_mean(square_diff)
    return loss


# creating the model
model = model()
model.summary()

# compile
adam = Adam(lr=1e-5, beta_1=0.9, beta_2=0.999, epsilon=1e-10, decay=0.0)
model.compile(optimizer=adam, loss=loss_function)

# train
epochs = 3 # kanav changed from 10 to 1
train_gen = train_generator(sample_per_batch=2, batch_number=15) # kanav changed from 1690 to 15
val_gen = valid_generator(sample_per_batch=5, batch_number=2) # kanav changed sample_per_batch from 50 to 5 & batch number from 10 to 2 

checkpoints = ModelCheckpoint('weights/weights{epoch:03d}.h5', save_weights_only=True, period=1)
history = model.fit_generator(train_gen, steps_per_epoch=2, epochs=epochs, verbose=1, shuffle=True,  # kanav changed from 1690 to 2
                              validation_data=val_gen, validation_steps=10,
                              callbacks=[checkpoints], max_queue_size=100)

with open('history.txt', 'a+') as f:
    print(history.history, file=f)

print('All Done!')


preprocess/generator.py
import os
import cv2
import pandas as pd
import random
import numpy as np
from visualize import visualize
from preprocess.datagen import label_generator, my_label_gen
from preprocess.augmentation import rotation, translate, crop, flip_horizontal, flip_vertical, noise, salt

# change the directories here when running on server/cloud - Kanav
train_directory = 'C:/Users/Kanav/Documents/Dissertation/Parkinsons_Disease/Codes/Fingertip-Mixed-Reality/Dataset/Train/' 
valid_directory = 'C:/Users/Kanav/Documents/Dissertation/Parkinsons_Disease/Codes/Fingertip-Mixed-Reality/Dataset/Valid/'
label_directory = 'C:/Users/Kanav/Documents/Dissertation/Parkinsons_Disease/Codes/Fingertip-Mixed-Reality/Dataset/label/'


def train_generator(sample_per_batch, batch_number):
    """ Generating training data """
    train_image_file = []

    image_files = os.listdir(train_directory)
    train_image_file = train_image_file + image_files
    print('Training Dataset Size: ' + str(len(train_image_file)))

    labels_data = pd.read_csv(label_directory + "combined_labels_v3_cleaned.csv")
    labels_data["filename"] = labels_data["video_name"] + "@" + labels_data.seq.str[3:6].astype(int).astype(str)
    
    for i in range(0, 10):
        random.seed(4)
        random.shuffle(train_image_file)

    while True:
        for i in range(0, batch_number - 1):
            start = i * sample_per_batch
            end = (i + 1) * sample_per_batch
            x_batch = []
            y_batch_key = []
            for n in range(start, end):
                image_name = train_image_file[n]
                
                # image: ndarray keypoints: ndarray but not normalized
                image = cv2.imread(train_directory + image_name)
                
                ## Resizing the image to a square dimension - Kanav
                old_size = image.shape[:2]
                desired_size = max(old_size)
                ratio = float(desired_size)/max(old_size)
                new_size = tuple([int(x*ratio) for x in old_size])
                new_size
                delta_w = desired_size - new_size[1]
                delta_h = desired_size - new_size[0]
                top, bottom = delta_h//2, delta_h-(delta_h//2)
                left, right = delta_w//2, delta_w-(delta_w//2)

                color = [0,0,0]
                image = cv2.copyMakeBorder(image, top, bottom, left, right, cv2.BORDER_CONSTANT,
                    value=color)
                
                #resizing image as per model requirements - change this when you try some other size - Kanav        
                image = cv2.resize(image, (128, 128), interpolation=cv2.INTER_LINEAR)
                
                # Initialize frame counter - kanav - may not be needed now
                #cnt = 0 
                filename = image_name.split(".jpg")[0]
                #file_index = int(image_name.split("@")[1].split(".")[0])
                #temp_labels = labels_data[labels_data["filename"] == filename]

                keypoints = []
                
                #adjusting the width labels as per new image dimensions of a square
                x1 = (labels_data[labels_data["filename"] == filename]["new_finger_x"]) + left
                y1 = (labels_data[labels_data["filename"] == filename]["new_finger_y"]) + top
                x2 = (labels_data[labels_data["filename"] == filename]["new_thumb_x"]) + left
                y2 = (labels_data[labels_data["filename"] == filename]["new_thumb_y"]) + top
                
                #adjusting labels as per dimension of reduced size for model
                x1 = (x1*128/(max(old_size)) )
                y1 = (y1*128/(max(old_size)) )
                x2 = (x2*128/(max(old_size)) )
                y2 = (y2*128/(max(old_size)) )

                keypoints.append(x1)
                keypoints.append(y1)
                keypoints.append(x2)
                keypoints.append(y2)

                    
#                 if 'TI1K' in image_name.split('_'):
#                     image, keypoints = my_label_gen(image_directory=train_directory,
#                                                     label_directory=label_directory,
#                                                     image_name=image_name)
#                 else:
#                     image, keypoints = label_generator(image_directory=train_directory,
#                                                        label_directory=label_directory,
#                                                        image_name=image_name)

                # 1.0 Original Image
                x_batch.append(image)
                y_batch_key.append(keypoints)
                # visualize(image, keypoints)

                """ Augmentation """
                # 2.0 Flip
                im_flip, k_flip = flip_horizontal(image, keypoints)
                x_batch.append(im_flip)
                y_batch_key.append(k_flip)
                # visualize(im_flip, k_flip)

                # 3.0 Original + rotation
                im, k = rotation(image, keypoints)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 4.0 Flip + rotation
                im, k = rotation(im_flip, k_flip)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 5.0 Original + translate
                im, k = translate(image, keypoints)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 6.0 Flip + translate
                im, k = translate(im_flip, k_flip)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 7.0 Original + crop
                im, k = crop(image, keypoints)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 8.0 Flip + crop
                im, k = crop(im_flip, k_flip)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 9.0 Original + noise
                im, k = noise(image, keypoints)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 10.0 Flip + noise
                im, k = noise(im_flip, k_flip)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 11.0 Original + salt
                im, k = salt(image, keypoints)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 12.0 Flip + salt
                im, k = salt(im_flip, k_flip)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 13.0 Original + flip vertical
                im, k = flip_vertical(image, keypoints)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 14.0 Flip + flip vertical
                im, k = flip_vertical(im_flip, k_flip)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 15.0 Original + rotate + translate
                im, k = rotation(image, keypoints)
                im, k = translate(im, k)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 16.0 Flip + rotate + translate
                im, k = rotation(im_flip, k_flip)
                im, k = translate(im, k)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 17.0 Original + noise + translate
                im, k = noise(image, keypoints)
                im, k = translate(im, k)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 18.0 Flip + noise + translate
                im, k = noise(im_flip, k_flip)
                im, k = translate(im, k)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 19.0 Original + crop + translate
                im, k = crop(image, keypoints)
                im, k = translate(im, k)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 20.0 Flip + crop + translate
                im, k = crop(im_flip, k_flip)
                im, k = translate(im, k)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

            x_batch = np.asarray(x_batch)
            x_batch = x_batch.astype('float32')
            x_batch = x_batch / 255.

            y_batch_key = np.asarray(y_batch_key)
            y_batch_key = y_batch_key.astype('float32')
            y_batch_key = y_batch_key / 128.
            yield (x_batch, y_batch_key)


def valid_generator(sample_per_batch, batch_number):
    """ Generating validation data """

    valid_image_files = os.listdir(valid_directory)
    labels_data = pd.read_csv(label_directory + "combined_labels_v3_cleaned.csv")
    labels_data["filename"] = labels_data["video_name"] + "@" + labels_data.seq.str[3:6].astype(int).astype(str)
    
    for i in range(0, 10):
        random.shuffle(valid_image_files)

    while True:
        for i in range(0, batch_number - 1):
            start = i * sample_per_batch
            end = (i + 1) * sample_per_batch
            x_batch = []
            y_batch_key = []
            for n in range(start, end):
                image_name = valid_image_files[n]
                
               # image: ndarray keypoints: ndarray but not normalized
                image = cv2.imread(valid_directory + image_name)
                
               ## Resizing the image to a square dimension - Kanav
                old_size = image.shape[:2]
                desired_size = max(old_size)
                ratio = float(desired_size)/max(old_size)
                new_size = tuple([int(x*ratio) for x in old_size])
                new_size
                delta_w = desired_size - new_size[1]
                delta_h = desired_size - new_size[0]
                top, bottom = delta_h//2, delta_h-(delta_h//2)
                left, right = delta_w//2, delta_w-(delta_w//2)

                color = [0,0,0]
                image = cv2.copyMakeBorder(image, top, bottom, left, right, cv2.BORDER_CONSTANT,
                    value=color)
                
                #resizing image as per model requirements - change this when you try some other size - Kanav        
                image = cv2.resize(image, (128, 128), interpolation=cv2.INTER_LINEAR)
                
                # Initialize frame counter - kanav - may not be needed now
                #cnt = 0 
                filename = image_name.split(".jpg")[0]
                #file_index = int(image_name.split("@")[1].split(".")[0])
                #temp_labels = labels_data[labels_data["filename"] == filename]
                keypoints = []
                
                #adjusting the width labels as per new image dimensions of a square
                x1 = (labels_data[labels_data["filename"] == filename]["new_finger_x"]) + left
                y1 = (labels_data[labels_data["filename"] == filename]["new_finger_y"]) + top
                x2 = (labels_data[labels_data["filename"] == filename]["new_thumb_x"]) + left
                y2 = (labels_data[labels_data["filename"] == filename]["new_thumb_y"]) + top
                
                #adjusting labels as per dimension of reduced size for model
                x1 = (x1*128/(max(old_size)) )
                y1 = (y1*128/(max(old_size)) )
                x2 = (x2*128/(max(old_size)) )
                y2 = (y2*128/(max(old_size)) )

                keypoints.append(x1)
                keypoints.append(y1)
                keypoints.append(x2)
                keypoints.append(y2)

                x_batch.append(image)
                y_batch_key.append(keypoints)

                im_flip, k_flip = flip_horizontal(image, keypoints)
                x_batch.append(im_flip)
                y_batch_key.append(k_flip)

            x_batch = np.asarray(x_batch)
            x_batch = x_batch.astype('float32')
            x_batch = x_batch / 255.

            y_batch_key = np.asarray(y_batch_key)
            y_batch_key = y_batch_key.astype('float32')
            y_batch_key = y_batch_key / 128.
            yield (x_batch, y_batch_key)


if __name__ == '__main__':
    train_directory = 'C:/Users/Kanav/Documents/Dissertation/Parkinsons_Disease/Codes/Fingertip-Mixed-Reality/Dataset/Train/' 
    valid_directory = 'C:/Users/Kanav/Documents/Dissertation/Parkinsons_Disease/Codes/Fingertip-Mixed-Reality/Dataset/Valid/'
    label_directory = 'C:/Users/Kanav/Documents/Dissertation/Parkinsons_Disease/Codes/Fingertip-Mixed-Reality/Dataset/label/'

    gen = train_generator(sample_per_batch=3, batch_number=2)
    x_batch, y_batch = next(gen)
    print(x_batch)
    print(y_batch)


preprocess/generator-orig.py
import os
import random
import numpy as np
from visualize import visualize
from preprocess.datagen import label_generator, my_label_gen
from preprocess.augmentation import rotation, translate, crop, flip_horizontal, flip_vertical, noise, salt

# change the directories here when running on server/cloud - Kanav
train_directory = 'C:/Users/Kanav/Documents/Dissertation/Parkinsons_Disease/Codes/Fingertip-Mixed-Reality/Dataset/Train/' 
valid_directory = 'C:/Users/Kanav/Documents/Dissertation/Parkinsons_Disease/Codes/Fingertip-Mixed-Reality/Dataset/Test/'
label_directory = 'C:/Users/Kanav/Documents/Dissertation/Parkinsons_Disease/Codes/Fingertip-Mixed-Reality/Dataset/label/'


def train_generator(sample_per_batch, batch_number):
    """ Generating training data """
    train_image_file = []

    image_files = os.listdir(train_directory)
    train_image_file = train_image_file + image_files
    print('Training Dataset Size: ' + str(len(train_image_file)))

    for i in range(0, 10):
        random.shuffle(train_image_file)

    while True:
        for i in range(0, batch_number - 1):
            start = i * sample_per_batch
            end = (i + 1) * sample_per_batch
            x_batch = []
            y_batch_key = []
            for n in range(start, end):
                image_name = train_image_file[n]

                # image: ndarray keypoints: ndarray but not normalized

                if 'TI1K' in image_name.split('_'):
                    image, keypoints = my_label_gen(image_directory=train_directory,
                                                    label_directory=label_directory,
                                                    image_name=image_name)
                else:
                    image, keypoints = label_generator(image_directory=train_directory,
                                                       label_directory=label_directory,
                                                       image_name=image_name)

                # 1.0 Original Image
                x_batch.append(image)
                y_batch_key.append(keypoints)
                # visualize(image, keypoints)

                """ Augmentation """
                # 2.0 Flip
                im_flip, k_flip = flip_horizontal(image, keypoints)
                x_batch.append(im_flip)
                y_batch_key.append(k_flip)
                # visualize(im_flip, k_flip)

                # 3.0 Original + rotation
                im, k = rotation(image, keypoints)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 4.0 Flip + rotation
                im, k = rotation(im_flip, k_flip)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 5.0 Original + translate
                im, k = translate(image, keypoints)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 6.0 Flip + translate
                im, k = translate(im_flip, k_flip)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 7.0 Original + crop
                im, k = crop(image, keypoints)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 8.0 Flip + crop
                im, k = crop(im_flip, k_flip)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 9.0 Original + noise
                im, k = noise(image, keypoints)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 10.0 Flip + noise
                im, k = noise(im_flip, k_flip)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 11.0 Original + salt
                im, k = salt(image, keypoints)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 12.0 Flip + salt
                im, k = salt(im_flip, k_flip)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 13.0 Original + flip vertical
                im, k = flip_vertical(image, keypoints)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 14.0 Flip + flip vertical
                im, k = flip_vertical(im_flip, k_flip)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 15.0 Original + rotate + translate
                im, k = rotation(image, keypoints)
                im, k = translate(im, k)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 16.0 Flip + rotate + translate
                im, k = rotation(im_flip, k_flip)
                im, k = translate(im, k)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 17.0 Original + noise + translate
                im, k = noise(image, keypoints)
                im, k = translate(im, k)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 18.0 Flip + noise + translate
                im, k = noise(im_flip, k_flip)
                im, k = translate(im, k)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 19.0 Original + crop + translate
                im, k = crop(image, keypoints)
                im, k = translate(im, k)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

                # 20.0 Flip + crop + translate
                im, k = crop(im_flip, k_flip)
                im, k = translate(im, k)
                x_batch.append(im)
                y_batch_key.append(k)
                # visualize(im, k)

            x_batch = np.asarray(x_batch)
            x_batch = x_batch.astype('float32')
            x_batch = x_batch / 255.

            y_batch_key = np.asarray(y_batch_key)
            y_batch_key = y_batch_key.astype('float32')
            y_batch_key = y_batch_key / 128.
            yield (x_batch, y_batch_key)


def valid_generator(sample_per_batch, batch_number):
    """ Generating validation data """

    valid_image_files = os.listdir(valid_directory)

    for i in range(0, 10):
        random.shuffle(valid_image_files)

    while True:
        for i in range(0, batch_number - 1):
            start = i * sample_per_batch
            end = (i + 1) * sample_per_batch
            x_batch = []
            y_batch_key = []
            for n in range(start, end):
                image_name = valid_image_files[n]

                # image: ndarray keypoints: ndarray but not normalized

                if 'TI1K' in image_name.split('_'):
                    image, keypoints = my_label_gen(image_directory=valid_directory,
                                                    label_directory=label_directory,
                                                    image_name=image_name)
                else:
                    image, keypoints = label_generator(image_directory=valid_directory,
                                                       label_directory=label_directory,
                                                       image_name=image_name)

                x_batch.append(image)
                y_batch_key.append(keypoints)

                im_flip, k_flip = flip_horizontal(image, keypoints)
                x_batch.append(im_flip)
                y_batch_key.append(k_flip)

            x_batch = np.asarray(x_batch)
            x_batch = x_batch.astype('float32')
            x_batch = x_batch / 255.

            y_batch_key = np.asarray(y_batch_key)
            y_batch_key = y_batch_key.astype('float32')
            y_batch_key = y_batch_key / 128.
            yield (x_batch, y_batch_key)


if __name__ == '__main__':
    train_directory = '../../Dataset/Train/'
    valid_directory = '../../Dataset/Valid/'
    label_directory = '../../Dataset/label/'

    gen = train_generator(sample_per_batch=10, batch_number=2)
    x_batch, y_batch = next(gen)
    print(x_batch)
    print(y_batch)


performance.py
### Updating the performance.py for my study
import os
import cv2
import time
import numpy as np
import pandas as pd
import imgaug as ia
from utils.iou import iou
from imgaug import augmenters as iaa
from yolo.detector import YOLO
from fingertip import Fingertips

#hand = YOLO('weights/yolo.h5', threshold=0.5)
fingertip = Fingertips(model='vgg', weights='weights/weights025.h5')
resolution = 1


# def flip_horizontal(img, keys):
#     """ Flipping """
#     aug = iaa.Sequential([iaa.Fliplr(1.0)])
#     seq_det = aug.to_deterministic()
#     keys = ia.KeypointsOnImage([ia.Keypoint(x=keys[0], y=keys[1]),
#                                 ia.Keypoint(x=keys[2], y=keys[3]),
#                                 ia.Keypoint(x=keys[4], y=keys[5]),
#                                 ia.Keypoint(x=keys[6], y=keys[7])], shape=img.shape)

#     image_aug = seq_det.augment_images([img])[0]
#     keys_aug = seq_det.augment_keypoints([keys])[0]
#     k = keys_aug.keypoints
#     keys_aug = [k[0].x, k[0].y, k[1].x, k[1].y, k[2].x, k[2].y, k[3].x, k[3].y]

#     return image_aug, keys_aug


image_directory = 'C:/Users/Kanav/Documents/Dissertation/Parkinsons_Disease/Codes/Fingertip-Mixed-Reality/Dataset/Valid/' 
label_directory = 'C:/Users/Kanav/Documents/Dissertation/Parkinsons_Disease/Codes/Fingertip-Mixed-Reality/Dataset/label/'
image_files = os.listdir(image_directory)

""" Ground truth label file for Parkinson's data """
labels_data = pd.read_csv(label_directory + "combined_labels_v3_cleaned.csv")
labels_data["filename"] = labels_data["video_name"] + "@" + labels_data.seq.str[3:6].astype(int).astype(str)

# file = open(label_directory + 'TI1K.txt')
# lines = file.readlines()
# file.close()

# """ Ground truth label file for SingleEight dataset """
# file = open(label_directory + 'SingleEight.txt')
# ego_lines = file.readlines()
# file.close()

total_error = np.zeros([1, 4])
avg_hand_detect_time = 0
avg_fingertip_detect_time = 0
avg_time = 0
avg_iou = 0
count = 0
distance_error = []

# height = int(480 * resolution)
# width = int(640 * resolution)

for image_name in image_files:
    """ Getting the ground truth labels """
    image = cv2.imread(image_directory + image_name)
    cols, rows, _ = image.shape
    if count == 0:
        print("row dim", rows)
        print("col dim", cols)
    filename = image_name.split(".jpg")[0]
    gt = []

    #fetching the groundtruth labels (based on cropped image of dim 780*910)
    xi = int(labels_data[labels_data["filename"] == filename]["new_finger_x"]) 
    yi = int(labels_data[labels_data["filename"] == filename]["new_finger_y"]) 
    xt = int(labels_data[labels_data["filename"] == filename]["new_thumb_x"]) 
    yt = int(labels_data[labels_data["filename"] == filename]["new_thumb_y"]) 
    

    
    gt = [xi, yi, xt, yt] #first index and then thumb
    
    """ Predictions for the test images """
    # Resizing the image to a square dimension - Kanav
    old_size = image.shape[:2]
    desired_size = max(old_size)
    ratio = float(desired_size)/max(old_size)
    new_size = tuple([int(x*ratio) for x in old_size])
    new_size
    delta_w = desired_size - new_size[1]
    delta_h = desired_size - new_size[0]
    top, bottom = delta_h//2, delta_h-(delta_h//2)
    left, right = delta_w//2, delta_w-(delta_w//2)

    color = [0,0,0]
    image = cv2.copyMakeBorder(image, top, bottom, left, right, cv2.BORDER_CONSTANT,
        value=color)

    #resizing image as per model requirements - change this when you try some other size - Kanav        
    image = cv2.resize(image, (128, 128), interpolation=cv2.INTER_LINEAR)

    tic3 = time.time()
    position = fingertip.classify(image=image)
    toc3 = time.time()
    avg_fingertip_detect_time = avg_fingertip_detect_time + (toc3 - tic3)

    for i in range(0, len(position), 2):
        position[i] = (position[i] * max(old_size)) - left
        position[i + 1] = (position[i + 1] * max(old_size)) - top

        # Commenting the below adjustment to keep labels within bounding box. But may use in future
#     for i in range(0, len(position), 2):
#         position[i] = (position[i] + tl[0])
#         position[i + 1] = (position[i + 1] + tl[1])

    pr = [position[0], position[1], position[2], position[3]]

    """ Calculating error for fingertips only """ 
    gt = np.asarray(gt)
    pr = np.asarray(pr)
    abs_err = abs(gt - pr)
    total_error = total_error + abs_err
    #D = np.sqrt((gt[0] - gt[2]) ** 2 + (gt[1] - gt[3]) ** 2)
    #D_hat = np.sqrt((pr[0] - pr[2]) ** 2 + (pr[1] - pr[3]) ** 2)
    #distance_error.append(abs(D - D_hat))
    count = count + 1
    print('Detected Image: {0}'.format(count))

er = total_error / count
#avg_iou = avg_iou / count
er = er[0]
er = np.round(er, 4)
# distance_error = np.array(distance_error)
# distance_error = np.mean(distance_error)
# distance_error = np.round(distance_error, 4)

print('Total Detected Image: {0}'.format(count))
#print('Average IOU: {0}'.format(avg_iou))
print('Pixel errors: xi = {0}, yi = {1}, xt = {2}, yt = {3}'.format(er[0], er[1],
                                                                                   er[2], er[3],
                                                                                   distance_error))

# avg_time = avg_time / 1000
# avg_hand_detect_time = avg_hand_detect_time / count
avg_fingertip_detect_time = avg_fingertip_detect_time / count

#print('Average execution time: {0:1.5f} ms'.format(avg_time * 1000))
#print('Average hand detection time: {0:1.5f} ms'.format(avg_hand_detect_time * 1000))
print('Average fingertip detection time: {0:1.5f} ms'.format(avg_fingertip_detect_time * 1000))

print('{0} & {1} & {2} & {3} '.format(er[0], er[1], er[2], er[3]))


